xwiki-pro-macros-api/src/main/java/com/xwiki/macros/AbstractProMacro.java

```diff
@@ -0,0 +1,133 @@
+ * @version $Id$
+ * @since 1.10.1
+ */
+@Unstable
+public abstract class AbstractProMacro<P> extends AbstractMacro<P>
+{
+    private static final LocalDocumentReference APP_WEBHOME = new LocalDocumentReference(Arrays.asList("Confluence",
+        "Macros"), "WebHome");
+
+    private static final LocalDocumentReference XWIKI_PREFERENCES = new LocalDocumentReference("XWiki",
+        "XWikiPreferences");
+
+    @Inject
+    private ContextualLocalizationManager localizationManager;
+
+    @Inject
+    private Licensor licensor;
+
+    @Inject
+    private DocumentAccessBridge accessBridge;
+
+    @Inject
+    private WikiDescriptorManager wikiDescriptorManager;
+
+    /**
+     * Creates a new macro instance.
+     *
+     * @param name the name of the macro.
+     * @param description the description of the macro.
+     * @param contentDescriptor the content descriptor of the macro.
+     * @param parametersBeanClass the class of the parameters bean of this class.
+     */
+    public AbstractProMacro(String name, String description,
+        ContentDescriptor contentDescriptor, Class<?> parametersBeanClass)
+    {
+        super(name, description, contentDescriptor, parametersBeanClass);
+    }
+
+    /**
+     * Creates a new macro instance.
+     *
+     * @param name the name of the macro.
+     * @param description the description of the macro.
+     * @param parametersBeanClass the class of the parameters bean of this class.
+     */
+    public AbstractProMacro(String name, String description, Class<?> parametersBeanClass)
+    {
+        super(name, description, parametersBeanClass);
+    }
+
+    /**
+     * Executes the macro.
+     *
+     * @param parameters the macro parameters in the form of a bean defined by the
+     *     {@link org.xwiki.rendering.macro.Macro} implementation
+     * @param content the content of the macro
+     * @param context the context of the macros transformation process
+     * @return the result of the macro execution as a list of Block elements
+     * @throws MacroExecutionException when an error occurs during the execution process
+     */
+    @Override
+    public List<Block> execute(P parameters, String content, MacroTransformationContext context)
+        throws MacroExecutionException
+    {
+        if (licensor.hasLicensure(
+            new DocumentReference(APP_WEBHOME, new WikiReference(wikiDescriptorManager.getCurrentWikiId()))))
+        {
+            return internalExecute(parameters, content, context);
+        } else {
+            String extensionName = localizationManager.getTranslationPlain("proMacros.extension.name");
+            String licenseAdminURL = accessBridge.getDocumentURL(
+                new DocumentReference(XWIKI_PREFERENCES, new WikiReference(wikiDescriptorManager.getMainWikiId())),
+                "admin", "editor=globaladmin&section=Licenses", "");
+            return Collections.singletonList(new MacroBlock(
```

Comment:
It's interesting that here you are returning directly an error macro block, while in ``NoteMacro`` you inject the warning macro and execute it. I'm not sure which option is best, but I'm wondering if you had any reason for doing it differently here and there? If there is a reason then maybe add a comment.

Reply:
For the note macro, I initially returned a new macro block, as I am doing with the error message. As a result, when editing with the WYSIWYG editor, I would get the warning message macro injected in my wiki syntax. From this `{{note}}text{{/note}}` to this `{{note}}{{warning}}text{{/warning}}{{/note}}`. Every time i would edit the page, one more warning macro was added. This was because when instantiating a new warning macro, I passed the current content of the note macro. I'll add a comment describing this.